x/ccv/provider/keeper/throttle.go

```diff
@@ -0,0 +1,496 @@
+//
+// Note: This queue is shared between pending slash packet data and pending vsc matured packet data.
+func (k Keeper) QueueThrottledSlashPacketData(
+	ctx sdktypes.Context, consumerChainID string, ibcSeqNum uint64, data ccvtypes.SlashPacketData) {
+	k.QueueThrottledPacketData(ctx, consumerChainID, ibcSeqNum, data)
+}
+
+// QueueThrottledVSCMaturedPacketData queues the vsc matured packet data for a chain-specific throttled packet data queue.
+//
+// Note: This queue is shared between pending slash packet data and pending vsc matured packet data.
+func (k Keeper) QueueThrottledVSCMaturedPacketData(
+	ctx sdktypes.Context, consumerChainID string, ibcSeqNum uint64, data ccvtypes.VSCMaturedPacketData) {
+	k.QueueThrottledPacketData(ctx, consumerChainID, ibcSeqNum, data)
+}
+
+// QueueThrottledPacketData queues a slash packet data or vsc matured packet data instance
+// for the given consumer chain's queue. This method is either used by tests, or called
+// by higher level methods with type assertion.
+func (k Keeper) QueueThrottledPacketData(
+	ctx sdktypes.Context, consumerChainID string, ibcSeqNum uint64, packetData interface{}) {
+
+	k.PanicIfTooMuchThrottledPacketData(ctx, consumerChainID)
+
+	store := ctx.KVStore(k.storeKey)
+
+	var bz []byte
+	var err error
+	switch data := packetData.(type) {
+	case ccvtypes.SlashPacketData:
+		bz, err = data.Marshal()
+		if err != nil {
+			panic(fmt.Sprintf("failed to marshal slash packet data: %v", err))
+		}
+		bz = append([]byte{slashPacketData}, bz...)
+	case ccvtypes.VSCMaturedPacketData:
+		bz, err = data.Marshal()
+		if err != nil {
+			panic(fmt.Sprintf("failed to marshal vsc matured packet data: %v", err))
+		}
+		bz = append([]byte{vscMaturedPacketData}, bz...)
+	default:
+		panic(fmt.Sprintf("unexpected packet data type: %T", data))
+	}
+
+	store.Set(providertypes.ThrottledPacketDataKey(consumerChainID, ibcSeqNum), bz)
+	k.IncrementThrottledPacketDataSize(ctx, consumerChainID)
+}
+
+// GetSlashAndTrailingData returns the first slash packet data instance and any
+// trailing vsc matured packet data instances in the chain-specific throttled packet data queue.
+//
+// Note: this method is not tested directly, but is covered indirectly
+// by TestHandlePacketDataForChain and e2e tests.
+func (k Keeper) GetSlashAndTrailingData(ctx sdktypes.Context, consumerChainID string) (
+	slashFound bool, slashData ccvtypes.SlashPacketData,
+	vscMaturedData []ccvtypes.VSCMaturedPacketData, ibcSeqNums []uint64) {
+
+	store := ctx.KVStore(k.storeKey)
+	iteratorPrefix := providertypes.ChainIdWithLenKey(providertypes.ThrottledPacketDataBytePrefix, consumerChainID)
+	iterator := sdktypes.KVStorePrefixIterator(store, iteratorPrefix)
+	defer iterator.Close()
+
+	slashFound = false
+	slashData = ccvtypes.SlashPacketData{}
+	vscMaturedData = []ccvtypes.VSCMaturedPacketData{}
+	ibcSeqNums = []uint64{}
+
+iteratorLoop:
```

Comment:
Change the switch to ifs to avoid using a label, becuase most people won't be familiar with labels

Reply:
See #611